From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <30385023+ishland@users.noreply.github.com>
Date: Tue, 16 Jun 2026 14:57:01 +0800
Subject: [PATCH] Buildscript setup


diff --git a/build.gradle b/build.gradle
index 74ab2006..0cc35632 100644
--- a/build.gradle
+++ b/build.gradle
@@ -361,6 +361,7 @@ dependencies {
 String getVersionSuffix() {
     def stdout = providers.exec {
         commandLine 'git', 'describe', '--tags', '--dirty', '--broken'
+        workingDir = project.file("..")
     }.standardOutput.asText.get()
     stdout = stdout.toString().strip()
     def suffix = ""
diff --git a/settings.gradle b/settings.gradle
index 12ca3b14..3634dbfd 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -10,7 +10,7 @@ pluginManagement {
 
 rootProject.name = "c2me" // avoid weirdness when building the project using another directory name
 
-includeBuild('FlowSched') {
+includeBuild('../FlowSched') {
     dependencySubstitution {
         substitute module('com.ishland.flowsched:flowsched') using project(':')
     }
