From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Thu, 16 Apr 2026 21:21:56 +0800
Subject: [PATCH] Buildscript setup


diff --git a/build.gradle b/build.gradle
index 4254df3f..b869b10e 100644
--- a/build.gradle
+++ b/build.gradle
@@ -319,6 +319,7 @@ dependencies {
 String getVersionSuffix() {
     def stdout = providers.exec {
         commandLine 'git', 'describe', '--tags', '--dirty', '--broken'
+        workingDir = project.file("..")
     }.standardOutput.asText.get()
     stdout = stdout.toString().strip()
     def suffix = ""
diff --git a/settings.gradle b/settings.gradle
index d3b65ab7..60070908 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -10,7 +10,7 @@ pluginManagement {
 
 rootProject.name = "c2me" // avoid weirdness when building the project using another directory name
 
-includeBuild('FlowSched') {
+includeBuild('../FlowSched') {
     dependencySubstitution {
         substitute module('com.ishland.flowsched:flowsched') using project(':')
     }
