From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <30385023+ishland@users.noreply.github.com>
Date: Sun, 21 Jun 2026 21:55:40 +0800
Subject: [PATCH] new: add
 dev.worldgen.tectonic.worldgen.densityfunction.Invert support


diff --git a/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/McToAst.java b/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/McToAst.java
index f5b12e19..410e098f 100644
--- a/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/McToAst.java
+++ b/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/McToAst.java
@@ -26,7 +26,6 @@ package com.ishland.c2me.opts.dfc.common.ast;
 
 import com.ishland.c2me.opts.dfc.common.Config;
 import com.ishland.c2me.opts.dfc.common.ast.binary.AddNode;
-import com.ishland.c2me.opts.dfc.common.ast.binary.DivNode;
 import com.ishland.c2me.opts.dfc.common.ast.binary.MaxNode;
 import com.ishland.c2me.opts.dfc.common.ast.binary.MaxShortNode;
 import com.ishland.c2me.opts.dfc.common.ast.binary.MinNode;
@@ -36,6 +35,7 @@ import com.ishland.c2me.opts.dfc.common.ast.conversion.ToF64Node;
 import com.ishland.c2me.opts.dfc.common.ast.integration.lithostitched.*;
 import com.ishland.c2me.opts.dfc.common.ast.integration.tectonic.ConfigClampBindings;
 import com.ishland.c2me.opts.dfc.common.ast.integration.tectonic.ConfigNoiseBindings;
+import com.ishland.c2me.opts.dfc.common.ast.integration.tectonic.InvertBindings;
 import com.ishland.c2me.opts.dfc.common.ast.misc.BeardifierNode;
 import com.ishland.c2me.opts.dfc.common.ast.misc.CacheLikeNode;
 import com.ishland.c2me.opts.dfc.common.ast.misc.ConstantF32Node;
@@ -207,6 +207,7 @@ public class McToAst {
             SqrtBindings.register(REGISTRY);
             ConfigClampBindings.register(REGISTRY);
             ConfigNoiseBindings.register(REGISTRY);
+            InvertBindings.register(REGISTRY);
         }
     }
 
diff --git a/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/integration/tectonic/InvertBindings.java b/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/integration/tectonic/InvertBindings.java
new file mode 100644
index 00000000..7bbc3ed0
--- /dev/null
+++ b/c2me-opts-dfc/src/main/java/com/ishland/c2me/opts/dfc/common/ast/integration/tectonic/InvertBindings.java
@@ -0,0 +1,87 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021-2026 ishland
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.ishland.c2me.opts.dfc.common.ast.integration.tectonic;
+
+import com.ishland.c2me.opts.dfc.common.ast.AstEmitter;
+import com.ishland.c2me.opts.dfc.common.ast.FrontendRegistry;
+import com.ishland.c2me.opts.dfc.common.ast.McToAst;
+import com.ishland.c2me.opts.dfc.common.ast.binary.DivNode;
+import com.ishland.c2me.opts.dfc.common.ast.misc.ConstantNode;
+import net.fabricmc.loader.api.FabricLoader;
+import net.minecraft.world.gen.densityfunction.DensityFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+
+public class InvertBindings {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(InvertBindings.class);
+
+    private static final Class<?> CLASS_Invert;
+    private static final MethodHandle MH_input;
+    private static final boolean AVAILABLE;
+
+    static {
+        Class<?> class_Invert = null;
+        MethodHandle mh_input = null;
+        boolean available = false;
+
+        if (FabricLoader.getInstance().isModLoaded("tectonic")) {
+            try {
+                class_Invert = Class.forName("dev.worldgen.tectonic.worldgen.densityfunction.Invert");
+                mh_input = MethodHandles.lookup().findVirtual(
+                        class_Invert,
+                        "input",
+                        MethodType.methodType(DensityFunction.class)
+                );
+
+                LOGGER.info("Bound to tectonic dev.worldgen.tectonic.worldgen.densityfunction.Invert");
+                available = true;
+            } catch (Throwable t) {
+                LOGGER.warn("Failed to bind to tectonic dev.worldgen.tectonic.worldgen.densityfunction.Invert", t);
+            }
+        }
+
+        CLASS_Invert = class_Invert;
+        MH_input = mh_input;
+        AVAILABLE = available;
+    }
+
+    public static void register(FrontendRegistry<AstEmitter<? extends DensityFunction>> registry) {
+        if (!AVAILABLE) return;
+
+        registry.registerExactMatch((Class<? extends DensityFunction>) CLASS_Invert, function -> {
+            try {
+                return new DivNode(new ConstantNode(1.0), McToAst.toAst((DensityFunction) MH_input.invoke(function)));
+            } catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+}
