From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <30385023+ishland@users.noreply.github.com>
Date: Thu, 18 Jun 2026 00:29:40 +0800
Subject: [PATCH] refactor: emulate ScopedValue for pre-J25


diff --git a/c2me-base/src/main/java/com/ishland/c2me/base/common/emulation/ScopedValueEmulation.java b/c2me-base/src/main/java/com/ishland/c2me/base/common/emulation/ScopedValueEmulation.java
new file mode 100644
index 00000000..7fa3bdfa
--- /dev/null
+++ b/c2me-base/src/main/java/com/ishland/c2me/base/common/emulation/ScopedValueEmulation.java
@@ -0,0 +1,413 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021-2026 ishland
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.ishland.c2me.base.common.emulation;
+
+import java.util.ArrayDeque;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * Note: This is an emulation of ScopedValue found on Java 25 with {@code ThreadLocal}
+ *
+ * <p> A value that may be safely and efficiently shared to methods without using method
+ * parameters.
+ *
+ * <p> In the Java programming language, data is usually passed to a method by means of a
+ * method parameter. The data may need to be passed through a sequence of many methods to
+ * get to the method that makes use of the data. Every method in the sequence of calls
+ * needs to declare the parameter and every method has access to the data.
+ * {@code ScopedValueEmulation} provides a means to pass data to a faraway method (typically a
+ * <em>callback</em>) without using method parameters. In effect, a {@code ScopedValueEmulation}
+ * is an <em>implicit method parameter</em>. It is "as if" every method in a sequence of
+ * calls has an additional parameter. None of the methods declare the parameter and only
+ * the methods that have access to the {@code ScopedValueEmulation} object can access its value
+ * (the data). {@code ScopedValueEmulation} makes it possible to securely pass data from a
+ * <em>caller</em> to a faraway <em>callee</em> through a sequence of intermediate methods
+ * that do not declare a parameter for the data and have no access to the data.
+ *
+ * <p> The {@code ScopedValueEmulation} API works by executing a method with a {@code ScopedValueEmulation}
+ * object <em>bound</em> to some value for the bounded period of execution of a method.
+ * The method may invoke another method, which in turn may invoke another. The unfolding
+ * execution of the methods define a <em>dynamic scope</em>. Code in these methods with
+ * access to the {@code ScopedValueEmulation} object may read its value. The {@code ScopedValueEmulation}
+ * object reverts to being <em>unbound</em> when the original method completes normally or
+ * with an exception. The {@code ScopedValueEmulation} API supports executing a {@link Runnable},
+ * or {@link CallableOp} with a {@code ScopedValueEmulation} bound to a value.
+ *
+ * <p> Consider the following example with a scoped value "{@code NAME}" bound to the value
+ * "{@code duke}" for the execution of a {@code Runnable}'s {@code run} method.
+ * The {@code run} method, in turn, invokes a method {@code doSomething}.
+ *
+ * {@snippet lang=java :
+ *     // @link substring="newInstance" target="#newInstance" :
+ *     private static final ScopedValueEmulation<String> NAME = ScopedValueEmulation.newInstance();
+ *
+ *     // @link substring="run" target="Carrier#run(Runnable)" :
+ *     ScopedValueEmulation.where(NAME, "duke").run(() -> doSomething());
+ * }
+ * Code executed directly or indirectly by {@code doSomething}, with access to the field
+ * {@code NAME}, can invoke {@code NAME.get()} to read the value "{@code duke}". {@code
+ * NAME} is bound while executing the {@code run} method. It reverts to being unbound when
+ * the {@code run} method completes.
+ *
+ * <p> The example using {@code run} invokes a method that does not return a result.
+ * The {@link Carrier#call(CallableOp) call} method can be used
+ * to invoke a method that returns a result.
+ * {@code ScopedValueEmulation} defines the {@link #where(ScopedValueEmulation, Object)} method
+ * for cases where multiple mappings (of {@code ScopedValueEmulation} to value) are accumulated
+ * in advance of calling a method with all {@code ScopedValueEmulation}s bound to their value.
+ *
+ * <h2>Bindings are per-thread</h2>
+ *
+ * A {@code ScopedValueEmulation} binding to a value is per-thread. Invoking {@code run}
+ * executes a method with a {@code ScopedValueEmulation} bound to a value for the current thread.
+ * The {@link #get() get} method returns the value bound for the current thread.
+ *
+ * <p> In the example, if code executed by one thread invokes this:
+ * {@snippet lang=java :
+ *     ScopedValueEmulation.where(NAME, "duke1").run(() -> doSomething());
+ * }
+ * and code executed by another thread invokes:
+ * {@snippet lang=java :
+ *     ScopedValueEmulation.where(NAME, "duke2").run(() -> doSomething());
+ * }
+ * then code in {@code doSomething} (or any method that it calls) invoking {@code NAME.get()}
+ * will read the value "{@code duke1}" or "{@code duke2}", depending on which thread is
+ * executing.
+ *
+ * <h2>Scoped values as capabilities</h2>
+ *
+ * A {@code ScopedValueEmulation} object should be treated as a <em>capability</em> or a key to
+ * access its value when the {@code ScopedValueEmulation} is bound. Secure usage depends on access
+ * control (see <cite>The Java Virtual Machine Specification</cite>, Section {@jvms 5.4.4})
+ * and taking care to not share the {@code ScopedValueEmulation} object. In many cases, a {@code
+ * ScopedValueEmulation} will be declared in a {@code final} and {@code static} field so that it
+ * is only accessible to code in a single class (or nest).
+ *
+ * <h2><a id="rebind">Rebinding</a></h2>
+ *
+ * The {@code ScopedValueEmulation} API allows a new binding to be established for <em>nested
+ * dynamic scopes</em>. This is known as <em>rebinding</em>. A {@code ScopedValueEmulation} that
+ * is bound to a value may be bound to a new value for the bounded execution of a new
+ * method. The unfolding execution of code executed by that method defines the nested
+ * dynamic scope. When the method completes, the value of the {@code ScopedValueEmulation} reverts
+ * to its previous value.
+ *
+ * <p> In the above example, suppose that code executed by {@code doSomething} binds
+ * {@code NAME} to a new value with:
+ * {@snippet lang=java :
+ *     ScopedValueEmulation.where(NAME, "duchess").run(() -> doMore());
+ * }
+ * Code executed directly or indirectly by {@code doMore()} that invokes {@code
+ * NAME.get()} will read the value "{@code duchess}". When {@code doMore()} completes
+ * then the value of {@code NAME} reverts to "{@code duke}".
+ *
+ * <p> Unless otherwise specified, passing a {@code null} argument to a method in this
+ * class will cause a {@link NullPointerException} to be thrown.
+ *
+ * @apiNote
+ * A {@code ScopedValueEmulation} should be preferred over a {@link ThreadLocal} for cases where
+ * the goal is "one-way transmission" of data without using method parameters.  While a
+ * {@code ThreadLocal} can be used to pass data to a method without using method parameters,
+ * it does suffer from a number of issues:
+ * <ol>
+ *   <li> {@code ThreadLocal} does not prevent code in a faraway callee from {@linkplain
+ *   ThreadLocal#set(Object) setting} a new value.
+ *   <li> A {@code ThreadLocal} has an unbounded lifetime and thus continues to have a value
+ *   after a method completes, unless explicitly {@linkplain ThreadLocal#remove() removed}.
+ *   <li> {@linkplain InheritableThreadLocal Inheritance} is expensive - the map of
+ *   thread-locals to values must be copied when creating each child thread.
+ * </ol>
+ *
+ * @param <T> the type of the value
+ */
+public class ScopedValueEmulation<T> {
+
+    private static final Object NIL = new Object();
+
+    private final ThreadLocal<ArrayDeque<Object>> backing = ThreadLocal.withInitial(ArrayDeque::new);
+
+    /**
+     * A mapping of scoped values, as <em>keys</em>, to values.
+     *
+     * <p> A {@code Carrier} is used to accumulate mappings so that an operation (a {@link
+     * Runnable} or {@link ScopedValueEmulation.CallableOp}) can be executed with all scoped values in the
+     * mapping bound to values. The following example runs an operation with {@code k1}
+     * bound (or rebound) to {@code v1}, and {@code k2} bound (or rebound) to {@code v2}.
+     * {@snippet lang=java :
+     *     // @link substring="where" target="#where(ScopedValue, Object)" :
+     *     ScopedValueEmulation.where(k1, v1).where(k2, v2).run(() -> ... );
+     * }
+     *
+     * <p> A {@code Carrier} is immutable and thread-safe. The {@link
+     * #where(ScopedValueEmulation, Object) where} method returns a new {@code Carrier} object,
+     * it does not mutate an existing mapping.
+     *
+     * <p> Unless otherwise specified, passing a {@code null} argument to a method in
+     * this class will cause a {@link NullPointerException} to be thrown.
+     *
+     */
+    public static final class Carrier {
+        final ScopedValueEmulation<?> key;
+        final Object value;
+        final Carrier prev;
+
+        Carrier(ScopedValueEmulation<?> key, Object value, Carrier prev) {
+            this.key = key;
+            this.value = value;
+            this.prev = prev;
+        }
+
+        /**
+         * Add a binding to this map, returning a new Carrier instance.
+         */
+        private static <T> Carrier where(ScopedValueEmulation<T> key, T value, Carrier prev) {
+            return new Carrier(key, value, prev);
+        }
+
+        /**
+         * Returns a new {@code Carrier} with the mappings from this carrier plus a
+         * new mapping from {@code key} to {@code value}. If this carrier already has a
+         * mapping for the scoped value {@code key} then it will map to the new
+         * {@code value}. The current carrier is immutable, so it is not changed by this
+         * method.
+         *
+         * @param key the {@code ScopedValue} key
+         * @param value the value, can be {@code null}
+         * @param <T> the type of the value
+         * @return a new {@code Carrier} with the mappings from this carrier plus the new mapping
+         */
+        public <T> Carrier where(ScopedValueEmulation<T> key, T value) {
+            return where(key, value, this);
+        }
+
+        /*
+         * Return a new set consisting of a single binding.
+         */
+        static <T> Carrier of(ScopedValueEmulation<T> key, T value) {
+            return where(key, value, null);
+        }
+
+        Object get() {
+            return value;
+        }
+
+        ScopedValueEmulation<?> getKey() {
+            return key;
+        }
+
+        private void pushValues0() {
+            for (Carrier carrier = this;
+                 carrier != null;
+                 carrier = carrier.prev) {
+                carrier.key.backing.get().push(wrapNullsafe(carrier.value));
+            }
+        }
+
+        private void popValues0() {
+            for (Carrier carrier = this;
+                 carrier != null;
+                 carrier = carrier.prev) {
+                ArrayDeque<Object> deque = carrier.key.backing.get();
+                Object peek = deque.getFirst();
+                if (unwrapNullsafe(peek) != carrier.value) {
+                    throw new AssertionError("Mismatch push and pop");
+                }
+                deque.pop();
+                if (deque.isEmpty()) {
+                    carrier.key.backing.remove();
+                }
+            }
+        }
+
+        /**
+         * Calls a value-returning operation with each scoped value in this mapping bound
+         * to its value in the current thread.
+         * When the operation completes (normally or with an exception), each scoped value
+         * in the mapping will revert to being unbound, or revert to its previous value
+         * when previously bound, in the current thread. If {@code op} completes with an
+         * exception then it propagated by this method.
+         *
+         * @param op the operation to run
+         * @param <R> the type of the result of the operation
+         * @param <X> type of the exception thrown by the operation
+         * @return the result
+         * @throws X if {@code op} completes with an exception
+         */
+        public <R, X extends Throwable> R call(ScopedValueEmulation.CallableOp<? extends R, X> op) throws X {
+            Objects.requireNonNull(op);
+            pushValues0();
+            try {
+                return op.call();
+            } finally {
+                popValues0();
+            }
+        }
+
+        /**
+         * Runs an operation with each scoped value in this mapping bound to its value
+         * in the current thread.
+         * When the operation completes (normally or with an exception), each scoped value
+         * in the mapping will revert to being unbound, or revert to its previous value
+         * when previously bound, in the current thread. If {@code op} completes with an
+         * exception then it propagated by this method.
+         *
+         * @param op the operation to run
+         */
+        public void run(Runnable op) {
+            Objects.requireNonNull(op);
+            pushValues0();
+            try {
+                op.run();
+            } finally {
+                popValues0();
+            }
+        }
+    }
+
+    /**
+     * An operation that returns a result and may throw an exception.
+     *
+     * @param <T> result type of the operation
+     * @param <X> type of the exception thrown by the operation
+     */
+    @FunctionalInterface
+    public interface CallableOp<T, X extends Throwable> {
+        /**
+         * Executes this operation.
+         * @return the result, can be null
+         * @throws X if the operation completes with an exception
+         */
+        T call() throws X;
+    }
+
+    /**
+     * Creates a new {@code Carrier} with a single mapping of a {@code ScopedValue}
+     * <em>key</em> to a value. The {@code Carrier} can be used to accumulate mappings so
+     * that an operation can be executed with all scoped values in the mapping bound to
+     * values. The following example runs an operation with {@code k1} bound (or rebound)
+     * to {@code v1}, and {@code k2} bound (or rebound) to {@code v2}.
+     * {@snippet lang=java :
+     *     // @link substring="run" target="Carrier#run(Runnable)" :
+     *     ScopedValue.where(k1, v1).where(k2, v2).run(() -> ... );
+     * }
+     *
+     * @param key the {@code ScopedValue} key
+     * @param value the value, can be {@code null}
+     * @param <T> the type of the value
+     * @return a new {@code Carrier} with a single mapping
+     */
+    public static <T> Carrier where(ScopedValueEmulation<T> key, T value) {
+        return Carrier.of(key, value);
+    }
+
+    /**
+     * Creates a scoped value that is initially unbound for all threads.
+     *
+     * @param <T> the type of the value
+     * @return a new {@code ScopedValue}
+     */
+    public static <T> ScopedValueEmulation<T> newInstance() {
+        return new ScopedValueEmulation<>();
+    }
+
+    /**
+     * {@return the value of the scoped value if bound in the current thread}
+     *
+     * @throws NoSuchElementException if the scoped value is not bound
+     */
+    @SuppressWarnings("unchecked")
+    public T get() {
+        ArrayDeque<Object> deque = this.backing.get();
+        if (deque.isEmpty()) {
+            this.backing.remove();
+            throw new NoSuchElementException();
+        }
+        return (T) unwrapNullsafe(deque.getFirst());
+    }
+
+    /**
+     * {@return {@code true} if this scoped value is bound in the current thread}
+     */
+    public boolean isBound() {
+        ArrayDeque<Object> deque = this.backing.get();
+        if (deque.isEmpty()) {
+            this.backing.remove();
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Returns the value of this scoped value if bound in the current thread, otherwise
+     * returns {@code other}.
+     *
+     * @param other the value to return if not bound
+     * @return the value of the scoped value if bound, otherwise {@code other}
+     */
+    @SuppressWarnings("unchecked")
+    public T orElse(T other) {
+        ArrayDeque<Object> deque = this.backing.get();
+        if (deque.isEmpty()) {
+            this.backing.remove();
+            return other;
+        } else {
+            return (T) unwrapNullsafe(deque.getFirst());
+        }
+    }
+
+    /**
+     * Returns the value of this scoped value if bound in the current thread, otherwise
+     * throws an exception produced by the exception supplying function.
+     *
+     * @param <X> the type of the exception that may be thrown
+     * @param exceptionSupplier the supplying function that produces the exception to throw
+     * @return the value of the scoped value if bound in the current thread
+     * @throws X if the scoped value is not bound in the current thread
+     */
+    @SuppressWarnings("unchecked")
+    public <X extends Throwable> T orElseThrow(Supplier<? extends X> exceptionSupplier) throws X {
+        Objects.requireNonNull(exceptionSupplier);
+        ArrayDeque<Object> deque = this.backing.get();
+        if (deque.isEmpty()) {
+            this.backing.remove();
+            throw exceptionSupplier.get();
+        } else {
+            return (T) unwrapNullsafe(deque.getFirst());
+        }
+    }
+
+    private static Object wrapNullsafe(Object value) {
+        return value != null ? value : NIL;
+    }
+
+    private static Object unwrapNullsafe(Object wrapped) {
+        Objects.requireNonNull(wrapped);
+        return wrapped == NIL ? null : wrapped;
+    }
+
+}
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationUtil.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationUtil.java
index f00a95aa..795e20a5 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationUtil.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationUtil.java
@@ -24,8 +24,10 @@
 
 package com.ishland.c2me.rewrites.chunksystem.common.async_chunkio;
 
+import com.ishland.c2me.base.common.emulation.ScopedValueEmulation;
+
 public class AsyncSerializationUtil {
 
-    public static final ScopedValue<Void> duringUnloadSerialization = ScopedValue.newInstance();
+    public static final ScopedValueEmulation<Void> duringUnloadSerialization = ScopedValueEmulation.newInstance();
 
 }
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
index 6d79ffcc..6b57ee59 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
@@ -27,6 +27,7 @@ package com.ishland.c2me.rewrites.chunksystem.common.statuses;
 import com.ibm.asyncutil.util.Either;
 import com.ishland.c2me.base.common.GlobalExecutors;
 import com.ishland.c2me.base.common.config.ModStatuses;
+import com.ishland.c2me.base.common.emulation.ScopedValueEmulation;
 import com.ishland.c2me.base.common.registry.SerializerAccess;
 import com.ishland.c2me.base.common.theinterface.IDirectStorage;
 import com.ishland.c2me.base.common.threadstate.ThreadInstrumentation;
@@ -255,7 +256,7 @@ public class ReadFromDisk extends NewChunkStatus {
 
             SerializedChunk serializer;
             if (Config.asyncSerialization) {
-                serializer = ScopedValue
+                serializer = ScopedValueEmulation
                         .where(AsyncSerializationUtil.duringUnloadSerialization, null)
                         .call(() -> SerializedChunk.fromChunk(((IThreadedAnvilChunkStorage) context.tacs()).getWorld(), chunk));
             } else {
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/structs/ChunkSystemExecutors.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/structs/ChunkSystemExecutors.java
index 3e646243..c80597fd 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/structs/ChunkSystemExecutors.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/structs/ChunkSystemExecutors.java
@@ -25,6 +25,7 @@
 package com.ishland.c2me.rewrites.chunksystem.common.structs;
 
 import com.ishland.c2me.base.common.GlobalExecutors;
+import com.ishland.c2me.base.common.emulation.ScopedValueEmulation;
 import io.reactivex.rxjava3.core.Scheduler;
 import io.reactivex.rxjava3.schedulers.Schedulers;
 
@@ -34,7 +35,7 @@ import java.util.concurrent.Executor;
 
 public class ChunkSystemExecutors {
 
-    public static final ScopedValue<Queue<Runnable>> CONSOLIDATING_QUEUE = ScopedValue.newInstance();
+    public static final ScopedValueEmulation<Queue<Runnable>> CONSOLIDATING_QUEUE = ScopedValueEmulation.newInstance();
 
     public static final Executor backingBackgroundExecutor = GlobalExecutors.prioritizedScheduler.executor(15);
     public static final Scheduler backgroundScheduler = Schedulers.from(backingBackgroundExecutor);
@@ -56,7 +57,7 @@ public class ChunkSystemExecutors {
 
     public static void consolidatingRoot0(ArrayDeque<Runnable> runnables) {
         backingBackgroundExecutor.execute(() -> {
-            ScopedValue
+            ScopedValueEmulation
                     .where(CONSOLIDATING_QUEUE, runnables)
                     .run(() -> {
                         while (!runnables.isEmpty()) {
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/mixin/MixinServerChunkManager.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/mixin/MixinServerChunkManager.java
index 85ce2fd2..56e77de8 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/mixin/MixinServerChunkManager.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/mixin/MixinServerChunkManager.java
@@ -24,6 +24,7 @@
 
 package com.ishland.c2me.rewrites.chunksystem.mixin;
 
+import com.ishland.c2me.base.common.emulation.ScopedValueEmulation;
 import com.ishland.c2me.base.mixin.access.IThreadedAnvilChunkStorage;
 import com.ishland.c2me.rewrites.chunksystem.common.ducks.IChunkSystemAccess;
 import com.ishland.c2me.rewrites.chunksystem.common.structs.ChunkSystemExecutors;
@@ -115,7 +116,7 @@ public abstract class MixinServerChunkManager {
     @WrapOperation(method = "updateChunks", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/world/ChunkLevelManager;update(Lnet/minecraft/server/world/ServerChunkLoadingManager;)Z"))
     private boolean consolidateSchedules(ChunkLevelManager instance, ServerChunkLoadingManager chunkLoadingManager, Operation<Boolean> original) {
         ArrayDeque<Runnable> runnables;
-        return ScopedValue
+        return ScopedValueEmulation
                 .where(ChunkSystemExecutors.CONSOLIDATING_QUEUE, runnables = new ArrayDeque<>())
                 .call(() -> {
                     try {
